### PR TITLE
Add logic and integration test for ResolveEntities.

### PR DESCRIPTION
**Pull request by Spaceenter**

### CODE REVIEW

File: test/integration/resolve_entities_test.go [hunk 1]
```diff
@@ -40,9 +41,38 @@ func TestResolveEntities(t *testing.T) {
 		req        *pb.ResolveEntitiesRequest
 		goldenFile string
 	}{
-		// TODO(spaceenter): Add real test case.
 		{
-			&pb.ResolveEntitiesRequest{},
+			&pb.ResolveEntitiesRequest{
```
Comment (shifucun) [inline test/integration/resolve_entities_test.go:45]: Would be nice to add an example curl command to the README doc

File: test/integration/resolve_entities_test.go [hunk 2]
```diff
@@ -40,9 +41,38 @@ func TestResolveEntities(t *testing.T) {
...
+				Entities: []*pb.EntitySubGraph{
+					{
+						SourceId: "newId/SantaClaraCountyId",
+						SubGraph: &pb.McfGraph{
+							Nodes: map[string]*pb.McfGraph_PropertyValues{
+								"newId/SantaClaraCountyId": {
+									Pvs: map[string]*pb.McfGraph_Values{
+										"wikidataId": {
+											TypedValues: []*pb.McfGraph_TypedValue{
```
Comment (shifucun) [inline test/integration/resolve_entities_test.go:54]: what are other options of "TypedValues"? It does not feel straightforward from API user point of view..

File: internal/proto/Debug.pb.go [hunk 3]
```diff
@@ -18,14 +18,13 @@
 
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.23.0
+// 	protoc-gen-go v1.26.0-devel
```
Comment (shifucun) [inline internal/proto/Debug.pb.go:21]: should the proto file be "debug.proto" instead of Debug.proto?

File: internal/util/util.go [hunk 4]
```diff
@@ -54,3 +56,12 @@ func UnzipAndDecode(contents string) ([]byte, error) {
...
 	return gzResult, nil
 }
+
+// Get the value of a given property, assuming single value.
+func GetPropVal(node *pb.McfGraph_PropertyValues, prop string) string {
+	values, ok := (node.GetPvs())[prop]
+	if !ok {
+		return ""
+	}
+	return (values.GetTypedValues())[0].GetValue()
```
Comment (shifucun) [inline internal/util/util.go:66]: check the length of values.GetTypedValues() >0 ?

File: internal/server/resolve.go [hunk 5]
```diff
@@ -2,30 +2,139 @@ package server
...
+	"github.com/datacommonsorg/reconciliation/internal/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+var (
+	// This is a preferred list.
+	// The props ranked higher are preferred over those ranked lower for resolving.
+	supportedIDProps = []string{
```
Comment (shifucun) [inline internal/server/resolve.go:19]: call this "rankedIDs"?

File: internal/server/resolve.go [hunk 6]
```diff
@@ -2,30 +2,139 @@ package server
...
-								Prop: "wikidataId",
-								Val:  "Q1234",
-							},
-						},
-						Probability: 0.58,
-					},
-				},
-			},
+	rowList := bigtable.RowList{}
+	idKey2SourceId := map[string]string{}
```
Comment (shifucun) [inline internal/server/resolve.go:35]: idKeyToSourceID

File: internal/server/resolve.go [hunk 7]
```diff
@@ -2,30 +2,139 @@ package server
...
+		if !ok {
+			continue
+		}
+
+		for _, idProp := range supportedIDProps {
+			idVal := util.GetPropVal(node, idProp)
+			if idVal == "" {
+				continue
+			}
+
```
Comment (shifucun) [inline internal/server/resolve.go:49]: can remove empty lines

File: internal/server/resolve.go [hunk 8]
```diff
@@ -2,30 +2,139 @@ package server
...
+			continue
+		}
+
+		parts := strings.Split(idKey, "^")
+		if len(parts) != 2 {
+			status.Errorf(codes.Internal, "Invalid id key %s", idKey)
+		}
+		idProp := parts[0]
+
+		if _, ok := sourceId2IdProp2ReconEntities[sourceId]; !ok {
```
Comment (shifucun) [inline internal/server/resolve.go:94]: maybe have simpler name for this...

File: test/integration/resolve_entities_test.go [hunk 9]
```diff
@@ -40,9 +41,38 @@ func TestResolveEntities(t *testing.T) {
...
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
```
Comment (shifucun) [inline test/integration/resolve_entities_test.go:75]: would be nice to add a few other test cases with unmatched result or multiple IDs

**Comment by shifucun** [comment]
There seems lint error from the "recon-pull-request" check

File: internal/proto/Debug.pb.go (hunk 3 above)
Comment (Spaceenter) [reply internal/proto/Debug.pb.go:21]: Debug.proto and Mcf.proto are borrowed/copied from the import repo (Java).

File: test/integration/resolve_entities_test.go (hunk 9 above)
Comment (Spaceenter) [reply test/integration/resolve_entities_test.go:124]: Done.

File: internal/server/resolve.go [hunk 10]
```diff
@@ -2,30 +2,140 @@ package server
...
+
+	// Collect to-be-resolved IDs to rowList and idKeyToSourceID.
+	for _, entity := range in.GetEntities() {
+		node, ok := (entity.GetSubGraph().GetNodes())[entity.GetSourceId()]
+		if !ok {
+			continue
+		}
+
+		for _, idProp := range rankedIDProps {
+			idVal := util.GetPropVal(node, idProp)
```
Comment (pradh) [inline internal/server/resolve.go:45]: maybe indicate in the interface Request proto that only a single value for the ID prop is considered?

File: proto/recon.proto [hunk 11]
```diff
@@ -21,6 +21,21 @@ package datacommons;
 import "google/api/annotations.proto";
 import "Mcf.proto";
 
+// Lists of entities (with their known IDs) for a given {idProp, idVal}.
+message ReconEntities {
```
Comment (pradh) [inline proto/recon.proto:25]: nit: If this is an internal proto, wonder if this should live in an internal proto? How does mixer handle this? (No change needed in this PR)

File: internal/server/resolve.go [hunk 12]
```diff
@@ -2,30 +2,140 @@ package server
...
+			if val, ok := idProp2ReconEntities[idProp]; ok {
+				reconEntities = val
+				break
+			}
+		}
+		if reconEntities == nil {
+			continue
+		}
+
+		probability := float64(1.0 / len(reconEntities.GetEntities()))
```
Comment (pradh) [inline internal/server/resolve.go:117]: Add a comment for why this is?

Comment (pradh) [reply internal/server/resolve.go:117]: A comment in proto definition of `probability` would help too.

File: test/integration/resolve_entities_test.go (hunk 2 above)
Comment (pradh) [reply test/integration/resolve_entities_test.go:55]: For reference-by-description this makes sense.  I slightly wonder if its useful to define a simpler interface for ID resolution?  We can bring it up in next meeting with Guha.

For instance, the response proto can what it is now, but the request can be either BySubGraph or ByIDValue, where the latter is a list of {sourceId, IdWithProperty}.

**Review by pradh — APPROVED** [review]
Thanks for the change!

File: test/integration/resolve_entities_test.go (hunk 2 above)
Comment (shifucun) [reply test/integration/resolve_entities_test.go:55]: Yea, I agree.. current form is not easy to use for client (especially someone new to our ontology or schema)

File: test/integration/resolve_entities_test.go [hunk 13]
```diff
@@ -40,9 +41,87 @@ func TestResolveEntities(t *testing.T) {
...
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+					// A new entity that cannot be resolved to a DC entity.
+					{
+						SourceId: "newId/MarsPlanetId",
```
Comment (shifucun) [inline test/integration/resolve_entities_test.go:77]: MarsPlanetId does not appear in result.json? should it have an empty entry?

**Review by shifucun — APPROVED** [review]

File: test/integration/resolve_entities_test.go (hunk 2 above)
Comment (Spaceenter) [reply test/integration/resolve_entities_test.go:55]: SG - will implement in a follow up PR.

File: proto/recon.proto (hunk 11 above)
Comment (Spaceenter) [reply proto/recon.proto:25]: Mixer has /proto/internal.proto, probably I can follow it. Noted this, will change in a follow up PR.

File: test/integration/resolve_entities_test.go (hunk 13 above)
Comment (Spaceenter) [reply test/integration/resolve_entities_test.go:77]: OK - added the non-resolved entities to result as empty struct.

File: internal/server/resolve.go (hunk 12 above)
Comment (Spaceenter) [reply internal/server/resolve.go:117]: Added comment. This is just for ID based resolving. Later on hopefully we have better computation of the probability.

File: internal/server/resolve.go (hunk 10 above)
Comment (Spaceenter) [reply internal/server/resolve.go:45]: Added comment here. Will add more comments to proto in a follow up CL when adding ByIDValue in addition to BySubGraph.

**Pull request merged**